src/data/nat/totient.lean

```diff
@@ -288,4 +289,40 @@ end
 @[simp] lemma totient_two : φ 2 = 1 :=
 (totient_prime prime_two).trans (by norm_num)
 
+theorem totient_Euler_product_formula (n : ℕ) :
+  φ n = n / (∏ p in n.factors.to_finset, p) * (∏ p in n.factors.to_finset, (p - 1)) :=
+begin
+  by_cases hn : n = 0, { subst hn, simp },
+  let P1 := ∏ p in n.factors.to_finset, p,
+  let P2 := ∏ p in n.factors.to_finset, (p - 1),
+  have h1 : 0 < P1,
+  { simpa [P1, ←factorization_rebase_prod] using prod_pos (λ p hp, pos_of_mem_factorization hp) },
+  suffices : φ n * P1 = n * P2,
```

Comment:
considering you prove this this way, I think it's good to have this form as a lemma too.

Reply:
I went ahead and did this and reorganized the proof, too:
```
theorem totient_formula_1 (n : ℕ) :
  φ n * ∏ p in n.factors.to_finset, p = n * ∏ p in n.factors.to_finset, (p - 1) :=
begin
  by_cases hn : n = 0, { simp [hn] },
  rw [multiplicative_factorization φ @totient_mul totient_one hn,
    ← factorization_rebase_prod, ← factorization_rebase_prod, ←finsupp.prod_mul],
  nth_rewrite 1 ←factorization_prod_pow_eq_self hn,
  rw ←finsupp.prod_mul,
  refine finsupp.prod_congr (λ p hp, _),
  have h : 0 < n.factorization p := zero_lt_iff.mpr (finsupp.mem_support_iff.mp hp),
  rw [totient_prime_pow (prime_of_mem_factorization hp) h,
    mul_comm, ←mul_assoc, ←pow_succ, nat.sub_add_cancel h],
end

theorem totient_formula_2 (n : ℕ) :
  φ n = n / (∏ p in n.factors.to_finset, p) * (∏ p in n.factors.to_finset, (p - 1)) :=
begin
  rw [← mul_div_left n.totient, totient_formula_1, mul_comm,
      nat.mul_div_assoc _ (prod_prime_factors_dvd n), mul_comm],
  simpa [factorization_rebase_prod] using prod_pos (λ p, pos_of_mem_factorization),
end
```

Reply:
Very nice, thanks!  I've golfed this a little further in a couple of places.

For want of a better idea, I'm calling the main formula `totient_eq_prod_div_prod`, and I've renamed your `totient_formula_1` as `totient_mul_prod_factors`.  This isn't a great name, but I can't think of a properly descriptive name that's not a short essay.

Reply:
sadly mathlib isn't averse to these essays - if you want some horrifying fun go look at the name for the law of cosines :')